pkg/jobs/job_scheduler.go

```diff
@@ -254,23 +288,40 @@ func (s *jobScheduler) executeSchedules(
 			continue
 		}
 
-		sp, err := txn.CreateSavepoint(ctx)
-		if err != nil {
-			return err
-		}
-
-		if err := s.processSchedule(ctx, schedule, numRunning, stats, txn); err != nil {
-			log.Errorf(ctx, "error processing schedule %d: %+v", schedule.ScheduleID(), err)
-
-			if err := txn.RollbackToSavepoint(ctx, sp); err != nil {
-				return errors.Wrapf(err, "failed to rollback savepoint for schedule %d", schedule.ScheduleID())
+		if processErr := withSavePoint(ctx, txn, func() error {
+			return s.processSchedule(ctx, schedule, numRunning, stats, txn)
+		}); processErr != nil {
+			if errors.HasType(err, (*savePointError)(nil)) {
+				return errors.Wrapf(err, "savepoint error for schedule %d", schedule.ScheduleID())
 			}
-		}
 
-		if err := txn.ReleaseSavepoint(ctx, sp); err != nil {
-			return err
+			// Failed to process schedule.
+			s.metrics.NumBadSchedules.Inc(1)
+			log.Errorf(ctx,
+				"error processing schedule %d: %+v", schedule.ScheduleID(), processErr)
+
+			// Try updating schedule record to indicate schedule execution error.
+			// We can't be sure that processSchedule didn't mutate schedule.
+			// So, we reload the schedule, and try to set schedule status.
+			if err := withSavePoint(ctx, txn, func() error {
+				schedule, _, err := s.unmarshalScheduledJob(row, cols)
+				if err != nil {
+					// This really shouldn't happen since we succeeded above.
+					return err
+				}
+				schedule.SetScheduleStatus(fmt.Sprintf("schedule planning error: %s", processErr))
```

Comment:
I might say `failed to create job: ` or something. "planning" is an internal concept and "schedule" is implied since we're in the schedule.